src/com/github/jhoenicke/javacup/ErrorManager.java

```diff
@@ -1,49 +1,113 @@
-
 package com.github.jhoenicke.javacup;
+
+import java.util.Set;
+import java.util.TreeSet;
+
 import com.github.jhoenicke.javacup.runtime.Symbol;
-public class ErrorManager{
-    private static ErrorManager errorManager;
-    private int errors = 0;
-    private int warnings = 0;
-    private int fatals = 0;
-    public int getFatalCount() { return fatals; }
-    public int getErrorCount() { return errors; }
-    public int getWarningCount() { return warnings; }
-    static {
-        errorManager = new ErrorManager();
-    }
-    public static ErrorManager getManager() { return errorManager; }
-    private ErrorManager(){
-    }
-
-    //TODO: migrate to java.util.logging
-    /**
-     * Error message format: 
-     * ERRORLEVEL at (LINE/COLUMN)@SYMBOL: MESSAGE
-     * ERRORLEVEL : MESSAGE
-     **/
-    public void emit_fatal(String message){
-        System.err.println("Fatal: "+message);
-        fatals++;
-    }
-    public void emit_fatal(String message, Symbol sym){
-        System.err.println("Fatal: "+message+" @ "+sym);
-        fatals++;
-    }
-    public void emit_warning(String message){
-        System.err.println("Warning: " + message);
-        warnings++;	
-    }
-    public void emit_warning(String message, Symbol sym){
-        System.err.println("Warning: " + message+" @ "+sym);
-        warnings++;
-    }
-    public void emit_error(String message){
-        System.err.println("Error: " + message);
-        errors++;
-    }
-    public void emit_error(String message, Symbol sym){
-        System.err.println("Error: "+message+" @ "+sym);
-        errors++;
-    }
+
+public class ErrorManager {
+
+	private static ErrorManager errorManager;
+	
+	static enum Severity {
+		Info,
+		Warning,
+		Error,
+		Fatal,
+	}
+
+	private int infos = 0;
+	private int warnings = 0;
+	private int errors = 0;
+	private int fatals = 0;
+
+	private Set<String> filter;
+	
+	private ErrorManager() {
+	}
+
+	public static ErrorManager getManager() {
+		if (errorManager == null) errorManager = new ErrorManager();
+		return errorManager;
+	}
+
+	public static void clear () {
+		errorManager = null;
+	}
+
+	private Set<String> getFilter () {
+		if (filter == null) filter = new TreeSet<>();
+		return filter;
+	}
+
+	public int getFatalCount() {
+		return fatals;
+	}
+
+	public int getErrorCount() {
+		return errors;
+	}
+
+	public int getWarningCount() {
+		return warnings;
+	}
+
+	/**
+	 * Error message format:
+	 *		ERRORLEVEL : MESSAGE @ Symbol: name#token=="value"(line/column)
+	 *		ERRORLEVEL : MESSAGE
+	 **/
+	
+	private void emit (Severity severity, String message, Symbol sym) {
+		if (sym != null && sym.value != null) {
```

Comment:
Sorry, I forgot to reply about the second run. No problem, In Main I call a clear method which nullify the instance so the getter will recreate one fresh with all fields set to initial value.

Reply:
I think this is a bit dangerous and sym.value may not even mean the right thing.  The symbol is really the location in the source file where the error happened and it can be any error like undeclared symbol, symbols declared multiple times, or a unrecoverable syntax error.  You may hide some error message here.

If you just don't want to get multiple warnings about undefined symbols, it would be better to do this in the parser and keep a set of symbols that were already warned about there.  This way you can be sure that only these kind of double reports are omitted and you get for free the clearing of the set.